### PR TITLE
Few controller specs added

### PR DESCRIPTION
**Pull request by nikolalsvk**

### CODE REVIEW

File: spec/controllers/char_attributes_controller.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+require 'rails_helper'
+
+describe CharAttributesController do  
+  describe "character attribute create" do
+
+    it "should create new character"
```
Comment (rastasheep) [inline spec/controllers/char_attributes_controller.rb]: You missed this one.

File: spec/controllers/char_types_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+      it "renders temaplte new" do
+        expect(response).to render_template("new")
+      end
+
+    end
+  end
+end
+
+
+# nesto.should_receive bla
```
Comment (pyromaniackeca) [inline spec/controllers/char_types_controller_spec.rb]: You can remove this now. xD

File: spec/controllers/char_attributes_controller.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
...
+
+
+  end
+end
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller.rb]: Add new line at the end of the file.

File: spec/controllers/char_attributes_controller_spec.rb [hunk 4]
```diff
@@ -0,0 +1,81 @@
+require 'rails_helper'
+
+describe CharAttributesController, :type => :controller do
+  # have legit user, char_type and char_attribute mocks
+  before do
+    @user = double(User)
+    @char_type = double(CharType)
+    @char_attribute = double(CharAttribute)
+
+    #allow(request.env["warden"]).to receive(:authenticate!).and_return(@user)
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: You can remove this line if it's not needed.

File: spec/controllers/char_attributes_controller_spec.rb [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+      end
+
+      it "should create new attribute" do
+        expect(assigns(:char_attribute)).to eql(@char_attribute)
+      end
+
+      it "should render template new" do
+        expect(response).to render_template("new")
+      end
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: I think you can test the response status here as well.

File: spec/controllers/char_types_controller_spec.rb [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      end
+
+      it "creates new character type" do
+        expect(assigns(:char_type)).to eql(@char_type)
+      end
+
+      it "renders temaplate new" do
+        expect(response).to render_template("new")
+      end
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_types_controller_spec.rb:130]: I think you can test the response status here as well.

File: spec/controllers/char_attributes_controller_spec.rb (hunk 5 above)
Comment (nikolalsvk) [inline spec/controllers/char_attributes_controller_spec.rb]: @pyromaniackeca where exaclty? before "expect(response).to render_template("new")" ?

Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: The exact order isn't important, but we prefer writing tests in the order they would happen in the code. under: `it "should render template new"` do ... end`
you can write `it should return http status success` do ... end`

File: spec/controllers/char_attributes_controller_spec.rb [hunk 7]
```diff
@@ -0,0 +1,186 @@
...
+    before do
+      find_char_type_stub
+      find_char_attribute_stub
+      get :show, valid_parameters
+    end
+
+    it "should have successful response" do
+      expect(response).to be_success
+    end
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: You can also test rendering here.

File: spec/controllers/char_attributes_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+      find_char_type_stub
+    end
+
+    it "should make a new character attribute" do
+      allow(@char_type).to receive_message_chain(:char_attributes, :new)
+      get :new, :char_type_id => 46
+
+      expect(response).to be_success
+    end
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb:56]: You can also test rendering here.

File: spec/controllers/char_attributes_controller_spec.rb [hunk 9]
```diff
@@ -0,0 +1,186 @@
...
+    before do
+      find_char_type_stub
+      find_char_attribute_stub
+      get :edit, valid_parameters
+    end
+
+    it "should have successful response" do
+      expect(response).to be_success
+    end
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: You can also test rendering here.

File: spec/controllers/char_attributes_controller_spec.rb [hunk 10]
```diff
@@ -0,0 +1,186 @@
...
+      end
+
+      it "should create new attribute" do
+        expect(assigns(:char_attribute)).to eql(@char_attribute)
+      end
+
+      it "should redirect to character type page" do
+        expect(response).to redirect_to(char_type_path(@char_type))
+      end
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: If you test redirecting status in one place, then you should test it in all places.

File: spec/controllers/char_attributes_controller_spec.rb [hunk 11]
```diff
@@ -0,0 +1,186 @@
...
+        put :update, :char_type_id => 46, :id => 2, 
+                     :char_attribute => { :title => "Rogue", :value => "20" }
+      end
+
+      it "redirects to character type page" do
+        expect(response).to redirect_to(char_type_path(@char_type))
+      end
+
+      it "should have http status redirect" do
+        expect(response).to have_http_status(302)
```
Comment (pyromaniackeca) [inline spec/controllers/char_attributes_controller_spec.rb]: Does rspec support a `be_redirect` macther? Or `have_http_status(:redirect)`?

File: spec/controllers/char_types_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+    @user = double(User)
+    @char_type = double(CharType)
+
+    allow(controller).to receive(:current_user).and_return(@user)
+    allow(controller).to receive(:authenticate_user!)
+
+  end
+
+  describe "#index" do
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_types_controller_spec.rb]: You can also test rendering here.

File: spec/controllers/char_types_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+      allow(@user).to receive_message_chain(:char_types,
+                                            :by_title, :page, :per).and_return(@char_type)
+      get :index, :page => "bla"
+
+      expect(response).to be_success
+    end
+  end
+
+  describe "#show" do
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_types_controller_spec.rb:24]: You can also test rendering here.

File: spec/controllers/char_types_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    it "should have successful response" do
+      find_char_type_stub
+      get :show, :id => 46
+
+      expect(response).to be_success
+    end
+  end
+
+  describe "#new" do
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_types_controller_spec.rb]: You can test rendering and http status here.

File: spec/controllers/char_types_controller_spec.rb [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+    it "should create new character type" do
+      allow(controller).to receive_message_chain(:current_user, :char_types, :new).and_return(@char_type)
+      get :new
+
+      expect(assigns(:char_type)).to eql(@char_type)
+    end
+  end
+
+  describe "#edit" do
+
```
Comment (pyromaniackeca) [inline spec/controllers/char_types_controller_spec.rb:62]: You can test rendering and assigning here.

**Comment by pyromaniackeca** [comment]
Generally everything looks good, but you're missing a lot of tests, I might have missed some, but you should test everything that happens in every action.

File: spec/controllers/shared_helpers_spec.rb [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+    it "should render 404 page" do
+      expect(response).to render_template("public/404")
+    end
+  end
+
+end
+
+shared_context "find character attribute" do
+
+end
```
Comment (pyromaniackeca) [inline spec/controllers/shared_helpers_spec.rb]: Add a new line at the end of the file.

File: spec/controllers/shared_helpers_spec.rb [hunk 17]
```diff
@@ -0,0 +1,42 @@
+shared_context "user logs in" do
+
+  before do
+    @user = double(User)
+    allow(controller).to receive(:current_user).and_return(@user)
+    allow(controller).to receive(:authenticate_user!)
+  end
+end
+
+shared_context "find character type" do
```
Comment (pyromaniackeca) [inline spec/controllers/shared_helpers_spec.rb]: I think you don't need most of this code now.

**Comment by rastasheep** [comment]
Beside that one comment by @pyromaniackeca when it becomes green on Semaphore, it's ready to be merged.

**Comment by pyromaniackeca** [comment]
:+1: It's good to merge.

**Pull request merged**